### PR TITLE
Adds support for Kong Workspaces

### PR DESCRIPTION
**Pull request by tjrivera**
Hi @hbagdi,

Starting small here but wanted to keep the ball rolling on getting more EE support in go-kong per #17.

This adds basic Workspace CRUD capabilities as well as a helper function to manage testing CE vs. EE.  

This does not include the management of Workspace Entities -- I'd like to keep PRs small if possible. 

Thanks,
Tyler

### CODE REVIEW

File: kong/workspace_service_test.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+	workspace, err = client.Workspaces.Get(defaultCtx, createdWorkspace.ID)
+	assert.Nil(err)
+	assert.NotNil(workspace)
+
+	workspace.Comment = String("new comment")
+	workspace, err = client.Workspaces.Update(defaultCtx, workspace)
+	assert.Nil(err)
+	assert.NotNil(workspace)
+	assert.Equal("teamA", *workspace.Name)
```
Comment (hbagdi) [inline kong/workspace_service_test.go:39]: Let's add an assert her to confirm `.Comment` is actually populated.

Can you also please add something in the `.Config` and `.Meta` to confirm they are correctly populated.

File: kong/workspace_service.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+	if err != nil {
+		return err
+	}
+
+	_, err = s.client.Do(ctx, req, nil)
+	return err
+}
+
+// List fetches a list of all Workspaces in Kong.
+func (s *WorkspaceService) List(ctx context.Context) ([]*Workspace, error) {
```
Comment (hbagdi) [inline kong/workspace_service.go]: Please match the signature of `List` method in other services. This should include ListOpt argument.

File: kong/workspace_service.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+		var workspace Workspace
+		err = json.Unmarshal(b, &workspace)
+		if err != nil {
+			return nil, err
+		}
+		workspaces = append(workspaces, &workspace)
+	}
+
+	return workspaces, nil
+}
```
Comment (hbagdi) [inline kong/workspace_service.go:149]: Please also add in `ListAll` method as well.

**Review by hbagdi — CHANGES_REQUESTED** [review]

**Comment by hbagdi** [comment]
Please rebase on latest master.

**Comment by hbagdi** [comment]
Tyler, can we please handle one endpoint at a time? Can you open a separate PR for workspace entity? That code requires more review it seems. The workspace code was almost ready to go but the entity work is blocking this PR.

File: kong/kong_test.go [hunk 4]
```diff
@@ -113,3 +121,22 @@ func runWhenKong(t *testing.T, semverRange string) {
 	}
 
 }
+
+// runWhenEnterprise skips a test if the version
+// of Kong running is not enterprise edition.
+func runWhenEnterprise(t *testing.T) {
```
Comment (hbagdi) [inline kong/kong_test.go]: This function should take a semver range as well. There are new endpoints introduced in newer versions of enterprise product that won't be supported by old versions and that will lead to test breakage.

File: kong/kong_test.go [hunk 5]
```diff
@@ -98,6 +98,14 @@ func runWhenKong(t *testing.T, semverRange string) {
...
 			v = v[:rcIndex]
 		}
+		eeIndex := strings.Index(v, "-enterprise-edition")
+		if eeIndex != -1 {
+			v = v[:eeIndex]
+		}
+		// A bit of a hack. Kong sometimes reports a version that
+		// does not match semver conventions. i.e. "1.5.0.1" will
+		// not parse properly.
+		v = strings.Join(strings.Split(v, ".")[:3], ".")
```
Comment (hbagdi) [inline kong/kong_test.go]: I'd recommend to implement this logic into a new method that can be re-used if need be. The previous logic was simple enough but with inclusion of this and the "rc" hack above that, we should probably refactor it. If you feel that's a scope creep to this PR and should rather be done separately, that's fine too.

File: kong/workspace_service_test.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+	createdWorkspace, err = client.Workspaces.Create(defaultCtx, workspace)
+	assert.Nil(err)
+	assert.NotNil(createdWorkspace)
+	assert.Equal(id, *createdWorkspace.ID)
+
+	err = client.Workspaces.Delete(defaultCtx, createdWorkspace.ID)
+	assert.Nil(err)
+}
+func TestWorkspaceServiceList(T *testing.T) {
```
Comment (hbagdi) [inline kong/workspace_service_test.go:62]: You don't have test for List()

File: kong/kong_test.go (hunk 4 above)
Comment (tjrivera) [reply kong/kong_test.go]: I've added support for semver range. thanks

File: kong/kong_test.go (hunk 5 above)
Comment (tjrivera) [reply kong/kong_test.go]: I think you're right, but I'm going to defer this refactor to a separate PR

File: kong/workspace_service_test.go (hunk 6 above)
Comment (tjrivera) [reply kong/workspace_service_test.go:62]: added. thanks!

**Comment by hbagdi** [comment]
There was a leftover WorkspaceEntity in `kong/types.go`. I removed that and merged in manually.
Thank you!
Let's move to the next one.
